rule/math_expr.go

```diff
@@ -0,0 +1,393 @@
+	return &exprMult{
+		operator: operator{
+			contract: Contract{
+				OpCode:     "mult",
+				ReturnType: NUMBER,
+				Terms: []Term{
+					{
+						Type:        NUMBER,
+						Cardinality: MANY,
+						Min:         2,
+					},
+				},
+			},
+		},
+	}
+}
+
+// Mult creates an expression that takes at least two operands, which
+// must evaluate to either Float64Value or Int64Value, and returns the
+// product of all operands.
+func Mult(vN ...Expr) Expr {
+	e := newExprMult()
+	e.consumeOperands(vN...)
+	return e
+}
+
+// Perform multiplication of Float64Value types.
+func (n *exprMult) float64Mult(params Params) (*Value, error) {
+	var product float64 = 1.0
+	for _, o := range n.operands {
+		f, err := exprToFloat64(o, params)
+		if err != nil {
+			return nil, err
+		}
+		product = product * f
+	}
+	return Float64Value(product), nil
+}
+
+// Perform multiplication of Int64Value types.
+func (n *exprMult) int64Mult(params Params) (*Value, error) {
+	var product int64 = 1
+	for _, o := range n.operands {
+		i, err := exprToInt64(o, params)
+		if err != nil {
+			return nil, err
+		}
+		product = product * i
+	}
+	return Int64Value(product), nil
+}
+
+// Eval makes exprMult comply with the Expr interface.
+func (n *exprMult) Eval(params Params) (*Value, error) {
+	// The ReturnType will be set to the concrete type that
+	// matches all the arguments by homogenisation.
+	if n.operator.Contract().ReturnType == FLOAT {
+		return n.float64Mult(params)
+	}
+	return n.int64Mult(params)
+}
+
+///////////////////
+// Div Operator //
+///////////////////
+type exprDiv struct {
+	operator
+}
+
+func newExprDiv() *exprDiv {
+	return &exprDiv{
+		operator: operator{
+			contract: Contract{
+				OpCode:     "div",
+				ReturnType: NUMBER,
+				Terms: []Term{
+					{
+						Type:        NUMBER,
+						Cardinality: MANY,
+						Min:         2,
+					},
+				},
+			},
+		},
+	}
+}
+
+// Div creates an expression that takes at least two operands, which
+// must evaluate to either Float64Value or Int64Value, and returns the
+// value of the first operand successively divided by the operands
+// that follow.
+func Div(vN ...Expr) Expr {
+	e := newExprDiv()
+	e.consumeOperands(vN...)
+	return e
+}
+
+// Perform division of Float64Value types.
+func (n *exprDiv) float64Div(params Params) (*Value, error) {
+	quotient, err := exprToFloat64(n.operands[0], params)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := 1; i < len(n.operands); i++ {
+		f, err := exprToFloat64(n.operands[i], params)
+		if err != nil {
+			return nil, err
+		}
+		quotient = quotient / f
+	}
+	return Float64Value(quotient), nil
+}
+
+// Perform division of Int64Value types.
+func (n *exprDiv) int64Div(params Params) (*Value, error) {
+	quotient, err := exprToInt64(n.operands[0], params)
+	if err != nil {
+		return nil, err
+	}
+
+	for j := 1; j < len(n.operands); j++ {
+		i, err := exprToInt64(n.operands[j], params)
+		if err != nil {
+			return nil, err
+		}
+		quotient = quotient / i
```

Comment:
Currently, if someone divides by 0, this line will panic.
This will translate into a 500 if the evaluation is done through HTTP, or a service crash if the evaluation is done client side and run in a separate goroutine.
This raises the question of how to deal with common errors relative to division by zero, overflow errors and other runtime errors: how do you think we should deal with those?

Reply:
Yes, I was thinking about this a little bit  - I've stopped short of doing errors and assertions about them here because the branch was already large.  But for the record, I do want to make assertions about error conditions.  The broader issue, as you say, is how we handle that at evaluation time.  We can't completely catch and eliminate those cases, and we don't yet having any branching operators that would allow the programmer to avoid them.  One idea is to have an ErrorValue and bubble this back up as the result of each evaluation, to be handled more "correctly" by the top level evaluator.  Obviously we can already return errors from "Eval" functions, but it might eventually be sensible to support some kind of error handling in the rules themselves, and having the `ErrorValue` would be a precursor for that.  

In an ideal world we wouldn't let any code that contained such errors reach production, but, again, without a way to handle it this will be hard.  We'd also need to think about how people can tests their rules without pushing them into production.